bot.py

```diff
@@ -700,11 +700,215 @@ def cmd_shoutout(match: Match[str]) -> Response:
+
+
+async def fetch_youtube_playlist_videos(
+        playlist_id: str,
+        *,
+        youtube_api_key: str
+) -> List[YouTubeVideo]:
+    url = 'https://www.googleapis.com/youtube/v3/playlistItems'
+    params = {
+        'part': 'snippet',
+        'playlistId': playlist_id,
+        'key': youtube_api_key,
+        'maxResults': 50,
+    }
+    playlist_videos = []
+    more_pages = True
+    async with aiohttp.ClientSession() as session:
+        while more_pages:
+            async with session.get(url, params=params) as resp:
+                json_resp = await resp.json()
+                playlist_videos.extend(json_resp.get('items'))
+                next_page_token = json_resp.get('nextPageToken')
+                if next_page_token:
+                    params['pageToken'] = next_page_token
+                else:
+                    more_pages = False
+
+        return [
+            YouTubeVideo(
+                video_id=v['snippet']['resourceId']['videoId'],
+                title=v['snippet']['title'],
+                playlist_id=v['snippet']['playlistId'],
+                # JSON response is zero-indexed unlike the UI
+                playlist_position=v['snippet']['position'] + 1,
+            ) for v in playlist_videos
+        ]
+
+
+async def add_youtube_videos_to_store(
+    db: aiosqlite.Connection,
+    videos: List[YouTubeVideo]
+) -> None:
+    for video in videos:
+        await db.execute(
+            (
+                'INSERT INTO youtube_playlist_videos ('
+                '    playlist_id, title, video_id, playlist_position'
+                ')'
+                'VALUES (?, ?, ?, ?)'
+            ),
+            (
+                video.playlist_id, video.title,
+                video.video_id, video.playlist_position,
+            ),
+        )
+        await db.commit()
+
+
+@async_lru.alru_cache(maxsize=30)
+async def populate_youtube_playlist_store(
+        playlist_id: str,
+        *,
+        youtube_api_key: str
+) -> None:
+    async with aiosqlite.connect('db.db') as db:
+        await ensure_youtube_playlist_videos_table_exists(db)
+        await clear_youtube_playlist_from_store(db, playlist_id)
+
+        videos = await fetch_youtube_playlist_videos(
+            playlist_id,
+            youtube_api_key=youtube_api_key,
+        )
+        await add_youtube_videos_to_store(db, videos)
+
+
+def youtube_video_row_factory(
+    cursor: aiosqlite.Cursor,
+    row: aiosqlite.Row,
+) -> YouTubeVideo:
+    d = {}
+    for idx, col in enumerate(cursor.description):
+        d[col[0]] = row[idx]
+    return YouTubeVideo(**d)
+
+
+async def search_youtube_videos(
+    db: aiosqlite.Connection,
+    playlist_id: str,
+    search_terms: str
+) -> List[YouTubeVideo]:
+    query = (
+        'SELECT playlist_id, title, video_id, playlist_position '
+        'FROM youtube_playlist_videos '
+        'WHERE playlist_id = ? AND title MATCH ? ORDER BY rank'
+    )
+    db.row_factory = youtube_video_row_factory
```

Comment:
whoa, I've never seen this before 🤔

Reply:
😄 

I'm guessing these 3 lines can be done with list comprehension but it's beyond my python skillz https://github.com/anthonywritescode/twitch-chat-bot/pull/30/files#diff-f289368d1ae6d2d58bec9f799832d932R820-R822